### PR TITLE
#3347 Fridge Detail Page

### LINKED ISSUE #3347: Sensor/Location Detail View
## Is your feature request related to a problem? Please describe.

The location/detail view as per the marvel prototype

## Describe the solution you'd like

The location/detail view as per the marvel prototype

## Implementation

TBD

## Describe alternatives you've considered

N/A

## Additional context

Resources

- https://marvelapp.com/prototype/55aea8f/screen/75648018
image

- https://drive.google.com/file/d/1woR4jOSMkfeHP_efsGrW5lU35pVmKvCR/view?usp=sharing
image

### PR DESCRIPTION
**Pull request by joshxg**
Fixes #3347 

## Change summary

- Adds `FridgeDetailPage`

## Testing

Internal

### Related areas to think about

image

- It's a bit slow. Will work on that in a new PR
- Bit big again. Soz. But! 200 lines for the astronaut!!
- Didn't do the header of the chart area yet, waiting to see what happens in the vaccine module page

### CODE REVIEW

File: src/database/DataTypes/TemperatureBreachConfiguration.js [hunk 1]
```diff
@@ -76,6 +76,7 @@ TemperatureBreachConfiguration.schema = {
     duration: { type: 'double', optional: true },
     description: { type: 'string', optional: true },
     colour: { type: 'string', optional: true },
+    location: { type: 'Location', optional: true },
```
Comment (joshxg) [inline src/database/DataTypes/TemperatureBreachConfiguration.js:79]: Will need to bump ya schema, to run this branch, probs

File: src/pages/FridgeDetailPage.js [hunk 2]
```diff
@@ -1 +1,223 @@
...
+  selectHotCumulativeBreach,
+  selectLeastRecentTemperatureLogDate,
+  selectMinAndMaxDomains,
+  selectMinAndMaxLogs,
+  selectMostRecentTemperatureLogDate,
+  selectNumberOfColdConsecutiveBreaches,
+  selectNumberOfHotConsecutiveBreaches,
+  selectTemperatureLogsFromDate,
+  selectTemperatureLogsToDate,
+} from '../selectors/fridge';
```
Comment (joshxg) [inline src/pages/FridgeDetailPage.js:36]: That's a lot of selectors.

I think as part of performance tuning (as it is a bit slow), looking at putting a wrapper around the chart which is connected to redux directly, as well as for the cards at the bottom, so the picking of state fields can be separated out into each of the two parts to be concerned about

File: src/pages/FridgeDetailPage.js [hunk 3]
```diff
@@ -1 +1,223 @@
...
+  const header = type.includes('CUMULATIVE') ? 'Cumulative Breach' : 'Consecutive Breach';
+
+  return (
+    <CardDetail
+      headerText={header}
+      Content={content ? <CardText color={color}>{content}</CardText> : <NoBreachMan />}
+      Footer={content ? <Icon color={color} /> : <CardText small>No breaches!</CardText>}
+    />
+  );
+};
```
Comment (joshxg) [inline src/pages/FridgeDetailPage.js:71]: Don't usually like having multiple components in the same file but also couldn't really justify taking them out.. I dunno

File: src/selectors/fridge.js [hunk 4]
```diff
@@ -39,6 +29,41 @@ export const selectFridgeTemperatureLogs = createSelector([selectSelectedFridge]
   return temperatureLogs.sorted?.('timestamp') ?? [];
 });
 
+export const selectLeastRecentTemperatureLogDate = createSelector(
```
Comment (joshxg) [inline src/selectors/fridge.js:32]: bunch of these selectors were made already - going to do a new issue and change them a bit to try and utilize realm queries more than memoizing selectors

File: src/pages/FridgeDetailPage.js [hunk 5]
```diff
@@ -1 +1,212 @@
...
+
+BreachCard.propTypes = {
+  content: PropTypes.oneOfType([PropTypes.string, PropTypes.object]).isRequired,
```
Comment (mark-prins) [inline src/pages/FridgeDetailPage.js:74]: `content` is required but can be null if there are no breaches

File: src/selectors/fridge.js [hunk 6]
```diff
@@ -129,3 +163,72 @@ export const selectTimestampFormatter = createSelector(
...
+  [selectSelectedFridge, selectFridgeTemperatureLogsFromDate],
+  (fridge, logs) => {
+    const { coldCumulativeBreachConfig } = fridge;
+
+    const { maximumTemperature, duration } = coldCumulativeBreachConfig;
+
+    const logsOverThreshold = logs.filtered('temperature <= $0', maximumTemperature);
+    const sum = logsOverThreshold.sum('logInterval') ?? 0;
+
+    return sum >= duration ? formatTime(sum * 1000) : null;
```
Comment (mark-prins) [inline src/selectors/fridge.js:219]: Is it worth taking into account the condition where `sum === 0` and `duration` is also `0` due to a bad configuration? the result is a display of a breach which is `0 m`
Technically correct, I suppose, as the duration is 0

**Review by mark-prins — APPROVED** [review]
LGTM!

Chart looks good 😄  even the no breach man is looking good.

File: src/database/DataTypes/Location.js [hunk 7]
```diff
@@ -57,6 +57,23 @@ export class Location extends Realm.Object {
   totalStock() {
     return this.batches.sum('numberOfPacks') ?? 0;
   }
+
+  get hotCumulativeBreachConfig() {
+    console.log('this.breachConfigs', this.breachConfigs);
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/database/DataTypes/Location.js]: 0ee058f

File: src/pages/FridgeDetailPage.js [hunk 8]
```diff
@@ -1 +1,212 @@
...
+  const color = type === 'HOT_CUMULATIVE' ? DANGER_RED : COLD_BREACH_BLUE;
+  const Icon = type === 'HOT_CUMULATIVE' ? HotBreachIcon : ColdBreachIcon;
+
+  const header = type.includes('CUMULATIVE') ? 'Cumulative Breach' : 'Consecutive Breach';
+
+  return (
+    <CardDetail
+      headerText={header}
+      Content={content ? <CardText color={color}>{content}</CardText> : <NoBreachMan />}
+      Footer={content ? <Icon color={color} /> : <CardText small>No breaches!</CardText>}
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/pages/FridgeDetailPage.js]: Whoops! 
f5a735a

[the remark this answers is not in the document]
Comment (joshxg) [reply src/pages/FridgeDetailPage.js]: abe7dcb

File: src/pages/FridgeDetailPage.js (hunk 5 above)
Comment (joshxg) [reply src/pages/FridgeDetailPage.js]: Hm I thought `object` type covered `null` also! good catch: 
47049c3

File: src/selectors/fridge.js (hunk 6 above)
Comment (joshxg) [reply src/selectors/fridge.js]: Yeah I think so. Could also consider having a minimum duration of (1), the same as the logging interval?

Comment (mark-prins) [reply src/selectors/fridge.js]: yeah, that's sensible. can't see why we should support a duration of `0`

**Pull request merged**